cpp/include/cudf/detail/structs/utilities.hpp

```diff
@@ -164,14 +164,18 @@ class flattened_table {
  * @param null_precedence null order for input table
  * @param nullability force output to have nullability columns even if input columns
  * are all valid
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate new device memory
  * @return `flatten_result` with flattened table, flattened column order, flattened null precedence,
  * alongside the supporting columns and device_buffers for the flattened table.
  */
 [[nodiscard]] flattened_table flatten_nested_columns(
   table_view const& input,
   std::vector<order> const& column_order,
   std::vector<null_order> const& null_precedence,
-  column_nullability nullability = column_nullability::MATCH_INCOMING);
+  column_nullability nullability,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

Comment:
Shouldn't we have no defaults in detail APIs, or is `mr` an exception?

Reply:
I thought about that too. Currently there is no guideline for this. I see that in https://github.com/rapidsai/cudf/pull/12888 @vyasr removed all the default value for `mr` although the original issue of that PR was asking for just removing default value for `stream`. So I think removing default value for this is still optional.

Reply:
Just FYI, @davidwendt created a series of PR removing default parameters for detail functions (e.g. #12003, #12006, #12007).

Reply:
These sound good. So I'm going to remove the default parameter completely in this related file.

Reply:
Oop, removing default `mr` value for the entire file would touch more places that are out of scope of this PR. I would rather do that in a separate PR.

Reply:
The relevant discussion on default parameters happened in #9584 and #11967. The tl;dr is that we want to remove default stream parameters in detail APIs, but there were concerns that this could lead to awkward parameter orderings (e.g. signatures like `..., stream, other_enum=enum_default, mr=rmm::mr::get_current_device_resource())`) if there are other default parameters. Since we don't know how many such APIs exist, the plan is to try to remove all default parameters and see how hard that is. In theory we could leave `mr` defaults since that parameter is always last, but the same arguments apply for `mr` as for `stream`: leaving that parameter defaulted leaves open room for developer error in forgetting to pass a user-provided `mr` from a public API to a detail one.

There is no official guideline yet because I am just now getting around to removing as many default parameters from detail APIs as is possible in PRs like #12888. If it turns out to be pretty painless then we should expect to update the dev docs with guidance to avoid default parameters altogether in detail APIs. As such, I'd go ahead and drop the mr default here. However, I agree with @ttnghia that the scope of this PR should be limited: don't add new default parameters, but also don't bother removing any that aren't directly related to your other changes.

Reply:
> don't add new default parameters, but also don't bother removing any that aren't directly related to your other changes.

:+1: